src/ansari/ansari_db.py

```diff
@@ -109,13 +112,95 @@ def _get_token_from_request(self, request: Request) -> str:
                 detail="Authorization header is malformed",
             )
 
+    def _execute_query(
+        self,
+        query: Union[str, list[str]],
+        params: Union[tuple, list[tuple]],
+        which_fetch: Union[Literal["one", "all"], list[Literal["one", "all"]]] = "",
+        commit_after: Literal["each", "all"] = "each",
+    ) -> Union[tuple, Optional[list[tuple]], list[Optional[list[tuple]]]]:
+        """
+        Executes one or more SQL queries with the provided parameters and fetch types.
+
+        Args:
+            query (Union[str, List[str]]): A single SQL query string or a list of SQL query strings.
+            params (Union[tuple, List[tuple]]): A single tuple of parameters or a list of tuples of parameters.
+            which_fetch (Union[Literal["one", "all"], List[Literal["one", "all"]]]):
+                A single fetch type or a list of fetch types. Each fetch type can be:
+                - "one": Fetch one row.
+                - "all": Fetch all rows.
+                - Any other value: Do not fetch any rows.
+            commit_after (Literal["each", "all"]): Whether to commit the transaction after each query is executed,
+                or only after all of them are executed.
+
+        Returns:
+            Union[Optional[List], List[Optional[List]]]:
+                - If a single query is executed:
+                    - Returns a single result if which_fetch is "one"
+                    - Returns a single list of rows if which_fetch is "all".
+                    - Else, returns None.
+                - If multiple queries are executed:
+                    - Returns a list of results, where each result is:
+                        - A single result if which_fetch is "one".
+                        - A list of rows if which_fetch is "all".
+                        - Else, returns None.
+
+            Note: the "single result" here could be a tuple if more than 1 column is selected in the query.
+
+        Raises:
+            ValueError: If an invalid fetch type is provided.
+        """
+        # If query is a single string, we assume that params and which_fetch are also non-list values
+        if isinstance(query, str):
+            requested_single_query = True
+            query = [query]
+            params = [params]
+            which_fetch = [which_fetch]
+        # else, we assume that params and which_fetch are lists of the same length
+        # and do a list-conversion just in case they are strings
+        else:
+            requested_single_query = False
+            if isinstance(params, str):
+                params = [params] * len(query)
+            if isinstance(which_fetch, str):
+                which_fetch = [which_fetch] * len(query)
+
+        caller_function_name = inspect.stack()[1].function
+        logger.debug(f"Function {caller_function_name}() \nis running queries: \n{query} \nwith params: \n{params}")
```

Comment:
What happens if this is called from an object method? Just curious.

Reply:
I don't quite get your question... if I understand correctly, an "object's method" is like `account_exists()` method in `Ansari` class (object) ... all other functions that are in this file are also methods which I call in it the `_execute_query() method. 

If you're asking what this code snippet does in general... then the answer is this: it fetches the name of the function/method that called this `_execute_query()` method (e.g., "account_exists", etc.)

Hopefully I understood your question correctly 🙌